runtime/sema/check_binary_expression.go

```diff
@@ -293,18 +297,29 @@ func (checker *Checker) checkBinaryExpressionArithmeticOrNonEqualityComparisonOr
 		)
 	}
 
-	switch operationKind {
-	case BinaryOperationKindArithmetic,
-		BinaryOperationKindBitwise:
-
-		return leftType
+	return leftType
+}
 
-	case BinaryOperationKindNonEqualityComparison:
-		return BoolType
+func (checker *Checker) checkBinaryExpressionNonEquality(
+	expression *ast.BinaryExpression,
+	operation ast.Operation,
+	leftType, rightType Type,
+	anyInvalid bool,
+) (resultType Type) {
+	resultType = BoolType
 
-	default:
-		panic(errors.NewUnreachableError())
+	if !(leftType.Equal(rightType) && leftType.IsComparable() && rightType.IsComparable()) && !anyInvalid {
```

Comment:
Maybe extract this condition to a util function (with early-exit pattern), just to make it a bit more readable..

Reply:
Refactored into a variable in 213c48038b8d24e076f22de3746a452f286c4a7e to improve readability